packages/perseus/src/widgets/interactive-graphs/reducer/interactive-graph-reducer.ts

```diff
@@ -359,24 +398,48 @@ function doChangeRange(
     };
 }
 
+const getDeltaVertex = (
+    maxMoves: vec.Vector2[],
+    minMoves: vec.Vector2[],
+    delta: vec.Vector2,
+): vec.Vector2 => {
+    const [deltaX, deltaY] = delta;
+    const maxXMove = Math.min(...maxMoves.map((move) => move[0]));
+    const maxYMove = Math.min(...maxMoves.map((move) => move[1]));
+    const minXMove = Math.max(...minMoves.map((move) => move[0]));
+    const minYMove = Math.max(...minMoves.map((move) => move[1]));
+    const dx = clamp(deltaX, minXMove, maxXMove);
+    const dy = clamp(deltaY, minYMove, maxYMove);
+    return [dx, dy];
+};
+
 const getChange = (
     coords: readonly vec.Vector2[],
     delta: vec.Vector2,
     constraintOpts: Omit<ConstraintArgs, "point">,
 ): vec.Vector2 => {
-    const [deltaX, deltaY] = delta;
     const maxMoves = coords.map((point: vec.Vector2) =>
         maxMove({...constraintOpts, point}),
     );
     const minMoves = coords.map((point: vec.Vector2) =>
         minMove({...constraintOpts, point}),
     );
-    const maxXMove = Math.min(...maxMoves.map((move) => move[0]));
-    const maxYMove = Math.min(...maxMoves.map((move) => move[1]));
-    const minXMove = Math.max(...minMoves.map((move) => move[0]));
-    const minYMove = Math.max(...minMoves.map((move) => move[1]));
-    const dx = clamp(deltaX, minXMove, maxXMove);
-    const dy = clamp(deltaY, minYMove, maxYMove);
+    const [dx, dy] = getDeltaVertex(maxMoves, minMoves, delta);
+    return [dx, dy];
+};
+
+const getNoSnapChange = (
+    coords: readonly vec.Vector2[],
+    delta: vec.Vector2,
+    constraintOpts: Omit<ConstraintArgs, "point" | "snapStep">,
+): vec.Vector2 => {
+    const maxMoves = coords.map((point: vec.Vector2) =>
+        maxNoSnapMove({...constraintOpts, point}),
+    );
+    const minMoves = coords.map((point: vec.Vector2) =>
+        minNoSnapMove({...constraintOpts, point}),
+    );
+    const [dx, dy] = getDeltaVertex(maxMoves, minMoves, delta);
```

Comment:
The near-duplication in all of these functions (`get(NoSnap)Change`, `(noSnap)bound`, `max(NoSnap)move`) makes me nervous. I think we could instead pass `snapStep: [0, 0]` to  the existing `getChange` function. This works because the snap step passed to `getChange` is only used to control how close a point can be to the edge of the graph — the actual snapping is done separately. Maybe the `snapStep` argument to `getChange` and its callees could be renamed to `minDistanceFromEdge` or `inset` or something like that to clarify how it's used.

Reply:
Totally no issue reverting that refactor. I just thought it was confusing, but it's not really worth it if it seems risky.

I'll work on updating the functions you mentioned. I didn't like the duplication either, but wasn't sure how to handle the snapStep. I'll try what you mentioned. Thanks! :)